[Preprocessing] Add pass to convert matmul to transposed variants

c-rhodes opened the pull request:
This introduces a pass that converts Linalg matmul ops to the transposed variants.

The pass is off by default and can be enabled with

    --iree-preprocessing-enable-transpose-matmul=<lhs, rhs>

c-rhodes committed:
[GlobalOptimization] Add pass to convert matmul to transposed variants

This introduces a pass that converts Linalg matmul ops to
the transposed variants.

The pass is off by default and can be enabled with

  --iree-global-opt-enable-transpose-matmul=<lhs, rhs>

MacDue commented on compiler/src/iree/compiler/GlobalOptimization/Passes.cpp:
```diff
@@ -100,6 +111,8 @@ void buildGlobalOptimizationPassPipeline(
       .addPass(createConvert1X1FilterConv2DToMatmulPass);
   mainPassManager.addPass(createEraseUnusedLinalgOperands());
 
+  mainPassManager.addPass(createTransposeMatmulPass(clTransposeMatmul));
```
I think it would be best to not create the pass if `clTransposeMatmul == TransposeMatmulOption::None`. 
I assume there's a cost to setting up and running a pass even if it does nothing.

c-rhodes committed:
dont add pass for none

MaheshRavishankar reviewed (changes requested):
Could we move this as a preprocessing pass? This is a very specific pass for SVE and is easier to manage as preprocessing pass instead of a global optimization pass with flags.

hanhanW reviewed (commented):
I think layout changes (driven by target backends) should be handled by data-tiling. This kind of pass is better to be in pre-processing phase.

c-rhodes committed:
move to preprocessing

MacDue commented on compiler/src/iree/compiler/Pipelines/Options.h:
```diff
@@ -167,9 +167,20 @@ struct SchedulingOptions {
   std::string preprocessingPassPipeline;
   std::string preprocessingTransformSpecFilename;
   std::string preprocessingPDLSpecFilename;
+  TransposeMatmulInput preprocessingTransposeMatmulInput;
```
Just going off other examples, it looks like you're intended to pass the pass preprocessing pass pipeline as a string, rather than add custom options for each pre-processing pass, e.g.:
```
--iree-preprocessing-pass-pipeline=builtin.module(func.func(iree-preprocessing-transpose-matmul{input=lhs}))
```

See: https://github.com/iree-org/iree/blob/main/tests/e2e/linalg/BUILD.bazel#L91

MacDue replied:
I also see there's a `iree-preprocessing-transpose-convolution-pipeline`, so if we want something more custom I feel adding `iree-preprocessing-transpose-matmul-pipeline` would be reasonable?

c-rhodes replied:
> I also see there's a iree-preprocessing-transpose-convolution-pipeline, so if we want something more custom I feel adding iree-preprocessing-transpose-matmul-pipeline would be reasonable?

I think so, it's certainly more convenient than the alternative.

hanhanW replied:
Putting a note: my understanding is that everything in preprocessing is for fast experiments. It will never be on by default. Otherwise we will move it to GlobalOptimization phase directly. When this pass becomes a need, we will need to revisit how to land it properly in IREE. IMO, the layout changes required by hardware targets can be served by data-tiling.

c-rhodes replied:
We're on the same page there, as I mentioned on the LLVM patch we're aware there's other options to investigate like data-tiling as you mention, but this gives us a big win for a small amount of effort. Going forward we'll be spending more time looking at the alternatives.

hanhanW commented on compiler/src/iree/compiler/Pipelines/Options.cpp:
```diff
@@ -245,6 +245,18 @@ void PreprocessingOptions::bindOptions(OptionsBinder &binder) {
       llvm::cl::cat(category));
+
+  binder.opt<TransposeMatmulInput>(
+      "iree-preprocessing-enable-transpose-matmul",
```
nit: I learned that `enable` is redundant in the flag -- when flags are boolean what decides if something is enabled or not is the value, not the presence of the flag, and it's just an extra word.

c-rhodes replied:
I initially tried that but it's the same as the flag in Passes.td and I hit a "flag is registered more than once" runtime error. That's why I added "enable", seems common in other flags as well. Open to suggestions on alternatives.

hanhanW replied:
oh I see, I did not notice that. How about add `-pass` suffix to the flag defined in `Passes.td` file?

MaheshRavishankar reviewed (approved):
Dont have a strong opinion on the pass. This looks OK to me, and I can live with it here. I would also point to https://github.com/iree-org/iree/pull/17144 which allows you to use PDLL (and there are similar flags for PDL/Transform dialect) that allow you to do this without an explicit pass.

hanhanW reviewed (approved):
Overall looks okay to me, I'll leave the details to @MacDue

c-rhodes committed:
rename passes

iree-preprocessing-transpose-matmul -> iree-preprocessing-transpose-matmul-pass
iree-preprocessing-enable-transpose-matmul -> iree-preprocessing-transpose-matmul

c-rhodes committed:
add transpose_matmul.mlir to Bazel + run clang-format

c-rhodes committed:
add TransposeMatmul.cpp to Bazel build and initialize option

c-rhodes committed:
add pipelines options to Bazel deps

c-rhodes committed:
add pipelines options to cmake deps

MaheshRavishankar reviewed (commented):
I think this shouldnt have landed this way. (Sorry I was on vacation). We should not need to introduce a new option. We should instead be using the `iree-preprocessing-pass-pipeline` flag. (https://github.com/MaheshRavishankar/iree/blob/e47f55dfccf6a9e3e4311dd6cef4346713e8bcdf/compiler/src/iree/compiler/Pipelines/Options.cpp#L225)

Could we revert this change and add this properly or just drop the pipeline options?

(cc @banach-space )
